includes/sanitizers/class-amp-meta-sanitizer.php

```diff
@@ -0,0 +1,333 @@
+		$width_found       = false;
+
+		foreach ( $viewport_settings as $index => $viewport_setting ) {
+			list( $property, $value ) = array_map( 'trim', explode( '=', $viewport_setting ) );
+			if ( 'width' === $property ) {
+				if ( 'device-width' !== $value ) {
+					$viewport_settings[ $index ] = 'width=device-width';
+				}
+				$width_found = true;
+				break;
+			}
+		}
+
+		if ( ! $width_found ) {
+			array_unshift( $viewport_settings, 'width=device-width' );
+		}
+
+		$viewport_tag->setAttribute( 'content', implode( ',', $viewport_settings ) );
+	}
+
+	/**
+	 * Parse and concatenate <amp-script> source meta tags.
+	 */
+	protected function process_amp_script_meta_tags() {
+		if ( empty( $this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ] ) ) {
+			return;
+		}
+
+		$first_meta_amp_script_src = array_shift( $this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ] );
+		$content_values            = [ $first_meta_amp_script_src->getAttribute( 'content' ) ];
+
+		// Merge (and remove) any subsequent meta amp-script-src elements.
+		while ( ! empty( $this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ] ) ) {
+			$meta_amp_script_src = array_shift( $this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ] );
+			$content_values[]    = $meta_amp_script_src->getAttribute( 'content' );
+		}
+
+		$first_meta_amp_script_src->setAttribute( 'content', implode( ' ', $content_values ) );
+
+		$this->meta_tags[ self::TAG_AMP_SCRIPT_SRC ][] = $first_meta_amp_script_src;
+	}
+
+	/**
+	 * Create a new meta tag for the charset value.
+	 *
+	 * @param string $charset Character set to use.
+	 * @return DOMElement New meta tag with requested charset.
+	 */
+	protected function create_charset_element( $charset ) {
+		return AMP_DOM_Utils::create_node(
+			$this->dom,
+			'meta',
+			[
+				'charset' => strtolower( $charset ),
+			]
+		);
+	}
+
+	/**
+	 * Create a new meta tag for the viewport setting.
+	 *
+	 * @param string $viewport Viewport setting to use.
+	 * @return DOMElement New meta tag with requested viewport setting.
+	 */
+	protected function create_viewport_element( $viewport ) {
+		return AMP_DOM_Utils::create_node(
+			$this->dom,
+			'meta',
+			[
+				'name'    => 'viewport',
+				'content' => $viewport,
+			]
+		);
+	}
+
+	/**
+	 * Check whether the charset is the correct one according to AMP requirements.
+	 *
+	 * @return bool Whether the charset is the correct one.
+	 * @throws LengthException If the charset meta tag was not previously retrieved.
+	 */
+	protected function is_correct_charset() {
+		if ( empty( $this->meta_tags[ self::TAG_CHARSET ] ) ) {
+			throw new LengthException( 'Failed to ensure a charset meta tag is present' );
+		}
+
+		$charset_element = $this->meta_tags[ self::TAG_CHARSET ][0];
+
+		return static::AMP_CHARSET === strtolower( $charset_element->getAttribute( 'charset' ) );
+	}
+
+	/**
+	 * Re-add the meta tags to the <head> node in the optimized order.
+	 *
+	 * The order is defined by the array entries in $this->meta_tags.
```

Comment:
Note the optimized order of DOM nodes is specified in https://amp.dev/documentation/guides-and-tutorials/optimize-and-measure/optimize_amp/

The comments in `\AMP_Theme_Support::ensure_required_markup()` are citing this document.

Reply:
Yes, I'll add an additional reference here. The comment above is meant to hint at the mechanism of how you can change the order that is being produced here.

Btw, the comments in the `AMP_Theme_Support::ensure_required_markup()` are the only good thing about procedural code... ;)